src/ng/parse.js

```diff
@@ -426,375 +574,1027 @@ Parser.prototype = {
+      right = this.recurse(ast.argument);
+      return this['unary' + ast.operator](right, context);
+    case AST.BinaryExpression:
+      left = this.recurse(ast.left);
+      right = this.recurse(ast.right);
+      return this['binary' + ast.operator](left, right, context);
+    case AST.LogicalExpression:
+      left = this.recurse(ast.left);
+      right = this.recurse(ast.right);
+      return this['binary' + ast.operator](left, right, context);
+    case AST.ConditionalExpression:
+      return this['ternary?:'](
+        this.recurse(ast.test),
+        this.recurse(ast.alternate),
+        this.recurse(ast.consequent),
+        context
+      );
+    case AST.Identifier:
+      ensureSafeMemberName(ast.name);
+      return function(scope, locals, assign, inputs) {
+        var base = locals && (ast.name in locals) ? locals : scope;
+        if (self.expensiveChecks || isPossiblyDangerousMemberName(ast.name)) {
+          ensureSafeObject(value, self.expression);
+        }
+        if (create && create !== 1 && base && !(ast.name in base)) {
+          base[ast.name] = {};
         }
-        var token = this.consume();
-        if (token.constant) {
-          keys.push(token.value);
-        } else if (token.identifier) {
-          keys.push(token.text);
+        var value = base ? base[ast.name] : undefined;
+        if (context) {
+          return {context: base, name: ast.name, value: value};
         } else {
-          this.throwError("invalid key", token);
+          return value;
         }
-        this.consume(':');
-        valueFns.push(this.expression());
-      } while (this.expect(','));
+      };
+    case AST.MemberExpression:
+      left = this.recurse(ast.object, false, !!create);
+      if (!ast.computed) {
+        ensureSafeMemberName(ast.property.name, self.expression);
+        right = ast.property.name;
+      }
+      if (ast.computed) right = this.recurse(ast.property);
+      return ast.computed ?
+        function(scope, locals, assign, inputs) {
+          var lhs = left(scope, locals, assign, inputs);
+          var rhs;
+          var value;
+          if (lhs != null) {
+            rhs = right(scope, locals, assign, inputs);
+            ensureSafeMemberName(rhs, self.expression);
+            if (create && create !== 1 && lhs && !(rhs in lhs)) {
+              lhs[rhs] = {};
+            }
+            value = lhs[rhs];
+            ensureSafeObject(value, self.expression);
+          }
+          if (context) {
+            return {context: lhs, name: rhs, value: value};
+          } else {
+            return value;
+          }
+        } :
+        function(scope, locals, assign, inputs) {
+          var lhs = left(scope, locals, assign, inputs);
+          if (create && create !== 1 && lhs && !(right in lhs)) {
+            lhs[right] = {};
+          }
+          var value = lhs != null ? lhs[right] : undefined;
+          if (self.expensiveChecks || isPossiblyDangerousMemberName(right)) {
```

Comment:
This `expensive || dangerous` check could be done once outside the closure. Not sure if that's worth it for CSP mode though? That would also remove the `self` reference though...